pkg/controller/direct/bigqueryconnection/connection_controller.go

```diff
@@ -223,26 +224,24 @@ func (a *Adapter) Update(ctx context.Context, updateOp *directbase.UpdateOperati
 	log := klog.FromContext(ctx).WithName(ctrlName)
 	log.V(2).Info("updating Connection", "name", a.id.External)
 	mapCtx := &direct.MapContext{}
-
-	updateMask := &fieldmaskpb.FieldMask{}
-	if !reflect.DeepEqual(a.desired.Spec.FriendlyName, a.actual.FriendlyName) {
-		updateMask.Paths = append(updateMask.Paths, "friendly_name")
-	}
-	if !reflect.DeepEqual(a.desired.Spec.Description, a.actual.Description) {
-		updateMask.Paths = append(updateMask.Paths, "description")
-	}
-
 	desired := a.desired.DeepCopy()
-	resource := BigQueryConnectionConnectionSpec_ToProto(mapCtx, &desired.Spec)
+	connection := BigQueryConnectionConnectionSpec_ToProto(mapCtx, &desired.Spec)
 	if mapCtx.Err() != nil {
 		return mapCtx.Err()
 	}
-
+	paths, err := common.CompareProtoMessage(connection, a.actual, common.BasicDiff)
+	if err != nil {
+		return err
+	}
+	if len(paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id.External)
+		return nil
+	}
 	fqn := a.id.External
 	req := &bigqueryconnectionpb.UpdateConnectionRequest{
 		Name:       fqn,
-		Connection: resource,
-		UpdateMask: updateMask,
+		Connection: connection,
+		UpdateMask: &fieldmaskpb.FieldMask{Paths: sets.List(paths)},
```

Comment:
Oh I like this ... I would expect the API to return an error here if we try to update an immutable field (or at least hope so), which makes it easier for us to write controllers!

Reply:
Ah OK.  I was actually with sending invalid requests to GCP if the user tells us to (i.e. we can tackle our backoff strategy)

If compareProtoMessage does all this maybe we give it a more important sounding name - something like "ConstructUpdateRequest"?